### PR TITLE
feat: Support for Vodafone Sercomm H300S

### PR DESCRIPTION
**Pull request by myhomeiot**
Tested with:
  Firmware: Vodafone-H-300s-v1.2.02.03
  Hardware: V2

Changed back to http://{host}/login.html used because with https Python 3.11+ return error: aiohttp.client_exceptions.ClientConnectorSSLError: Cannot connect to host 192.168.10.1:443 ssl:False [[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1006)]

processEnterFunc with challenge login: [login.js.txt](https://github.com/chemelli74/aiovodafone/files/14268014/login.js.txt)

### CODE REVIEW

File: src/aiovodafone/api.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ async def get_device_type(
                 if "data" in response_json and "ModelName" in response_json["data"]:
                     return DeviceType.TECHNICOLOR
         async with session.get(
-            f"https://{host}/login.html", headers=HEADERS, ssl=False
+            f"http://{host}/login.html", headers=HEADERS, ssl=False
```
Comment (chemelli74) [inline src/aiovodafone/api.py:69]: Sercomm devices work correctly only with https

File: src/aiovodafone/api.py [hunk 2]
```diff
@@ -414,23 +414,39 @@ async def _reset(self) -> bool:
...
+        self, username: str, password: str, useChallenge: bool = False
+    ) -> bool:
         """Login via json page"""
 
-        payload = {
-            "LoginName": username,
-            "LoginPWD": password,
-        }
+        if useChallenge:
+            return_dict = await self._get_sercomm_page("/data/login.json")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:423]: Please create a method that return the correct payload to make it easier to read

File: src/aiovodafone/api.py [hunk 3]
```diff
@@ -414,23 +414,39 @@ async def _reset(self) -> bool:
...
+                "LoginPWD": password,
+            }
+
         reply_json = await self._post_sercomm_page("/data/login.json", payload)
-        _LOGGER.debug("Login result: %s[%s]", LOGIN[int(str(reply_json))], reply_json)
+        reply_int = int(str(reply_json))
+        _LOGGER.debug("Login result: %s[%s]", LOGIN[reply_int], reply_json)
 
-        if reply_json == "1":
+        if reply_int == 1:
```
Comment (chemelli74) [inline src/aiovodafone/api.py:443]: This won't work with Vodafone WiFi 6 Station (RHG3006)

File: src/aiovodafone/api.py [hunk 4]
```diff
@@ -526,19 +542,25 @@ async def login(self) -> bool:
...
-                await self._encrypt_string(self.password),
-            )
-        except CannotAuthenticate:
-            _LOGGER.debug("Login second try: username[plain], password[encrypted]")
-            logged = await self._login_json(
-                self.username, await self._encrypt_string(self.password)
-            )
+        if not self.encryption_key:
+            _LOGGER.debug("Login: username[plain], password[challenge encrypted]")
+            logged = await self._login_json(self.username, self.password, True)
```
Comment (chemelli74) [inline src/aiovodafone/api.py:547]: you can leave the code as it is and use "self.encryption_key" as third parameter of the _login_json instead

**Review by chemelli74 — CHANGES_REQUESTED** [review]

File: src/aiovodafone/api.py (hunk 3 above)
Comment (myhomeiot) [reply src/aiovodafone/api.py:432]: Original code [use](https://github.com/chemelli74/aiovodafone/blob/85041a1c26fe917a2c0a843c12e9502ed26b4d60/src/aiovodafone/api.py#L426) LOGIN[int(str(reply_json))], I just make reply_int = int(str(reply_json)) from it and use.

**Comment by chemelli74** [comment]
@myhomeiot, are you going to update the code so we can consider merging ?

**Comment by myhomeiot** [comment]
@chemelli74 I resolve the conflicts and check resulting code with HA 2024.4.3, working as expected.
PS: I don't understand how to fix the lint errors without delete this pull request and reopen it. If you can, please fix this strange lint errors.
Thanks in advance.

**Comment by chemelli74** [comment]
You need to use pre-commit ;-)

image

**Comment by chemelli74** [comment]
> How you want to fix PLR2004 error (Magic value used in comparison), should I add this constants into const.py or make an enum?

I think the cleanest approach is to transform LOGIN from const.py in a dict and use values from there.

**Comment by chemelli74** [comment]
You are committing to your `main` branch so I cannot co-author.
Please use always pre-commit in future.

**Comment by myhomeiot** [comment]
Fix CI problems with new pull request.

**Pull request closed without merging**